File: app/src/main/java/com/imdc/milkdespencer/common/SharedPreferencesManager.java
package com.imdc.milkdespencer.common;

import android.content.Context;
import android.content.SharedPreferences;

public class SharedPreferencesManager {

    private static final String PREFERENCES_NAME = "YourSharedPreferencesName";
    private static SharedPreferencesManager instance;
    private final SharedPreferences sharedPreferences;
    private final SharedPreferences.Editor editor;

    private SharedPreferencesManager(Context context) {
        sharedPreferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE);
        editor = sharedPreferences.edit();
    }

    public static synchronized SharedPreferencesManager getInstance(Context context) {
        if (instance == null) {
            instance = new SharedPreferencesManager(context);
        }
        return instance;
    }



    public boolean hasValue(String key) {
        return sharedPreferences.contains(key);
    }

    public void save(String key, Object value) {
        if (value instanceof String) {
            editor.putString(key, (String) value);
        } else if (value instanceof Float) {
            editor.putFloat(key, (Float) value);
        } else if (value instanceof Integer) {
            editor.putInt(key, (Integer) value);
        } else if (value instanceof Boolean) {
            editor.putBoolean(key, (Boolean) value);
        } // Add more cases for other data types as needed

        editor.apply();
    }
    public Object get(String key, Object defaultValue) {
        if (defaultValue instanceof String) {
            return sharedPreferences.getString(key, (String) defaultValue);
        } else if (defaultValue instanceof Float) {
            return sharedPreferences.getFloat(key, (Float) defaultValue);
        } else if (defaultValue instanceof Integer) {
            return sharedPreferences.getInt(key, (Integer) defaultValue);
        } else if (defaultValue instanceof Boolean) {
            return sharedPreferences.getBoolean(key, (Boolean) defaultValue);
        }

        return defaultValue;
    }


    /*Delete from shared preference*/
    public void delete(String key) {
        if (sharedPreferences.contains(key)) {
            editor.remove(key);
            editor.apply();
        }
    }


    public void saveUsername(String username) {
        save("username", username);
    }

    public String getUsername() {
        return (String) get("username", "");
    }

    public void saveUserId(int userId) {
        save("user_id", userId);
    }

    public int getUserId() {
        return (int) get("user_id", -1);
    }
}

File: app/src/main/java/com/imdc/milkdespencer/common/SharedPreferencesManager.kt
package com.imdc.milkdespencer.common

import android.content.Context
import android.content.SharedPreferences

class SharedPreferencesManager private constructor(context: Context) {
    private val sharedPreferences: SharedPreferences
    private val editor: SharedPreferences.Editor

    init {
        sharedPreferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
        editor = sharedPreferences.edit()
    }

    fun hasValue(key: String?): Boolean {
        return sharedPreferences.contains(key)
    }

    fun save(key: String?, value: Any?) {
        if (value is String) {
            editor.putString(key, value as String?)
        } else if (value is Float) {
            editor.putFloat(key, (value as Float?)!!)
        } else if (value is Int) {
            editor.putInt(key, (value as Int?)!!)
        } else if (value is Boolean) {
            editor.putBoolean(key, (value as Boolean?)!!)
        } // Add more cases for other data types as needed
        editor.apply()
    }

    operator fun get(key: String?, defaultValue: Any?): Any? {
        if (defaultValue is String) {
            return sharedPreferences.getString(key, defaultValue as String?)
        } else if (defaultValue is Float) {
            return sharedPreferences.getFloat(key, (defaultValue as Float?)!!)
        } else if (defaultValue is Int) {
            return sharedPreferences.getInt(key, (defaultValue as Int?)!!)
        } else if (defaultValue is Boolean) {
            return sharedPreferences.getBoolean(key, (defaultValue as Boolean?)!!)
        }
        return defaultValue
    }

    /*Delete from shared preference*/
    fun delete(key: String?) {
        if (sharedPreferences.contains(key)) {
            editor.remove(key)
            editor.apply()
        }
    }

    fun saveUsername(username: String?) {
        save("username", username)
    }

    val username: String?
        get() = get("username", "") as String?

    fun saveUserId(userId: Int) {
        save("user_id", userId)
    }

    val userId: Int
        get() = get("user_id", -1) as Int

    companion object {
        private const val PREFERENCES_NAME = "YourSharedPreferencesName"
        private var instance: SharedPreferencesManager? = null
        @JvmStatic
        @Synchronized
        fun getInstance(context: Context): SharedPreferencesManager? {
            if (instance == null) {
                instance = SharedPreferencesManager(context)
            }
            return instance
        }
    }
}
